### DISCUSSION TITLE
Support prop prefetching when using SSR

### DISCUSSION THREAD

**Discussion by nihakue** [vercel/next.js#26415] [Ideas]
### Describe the feature you'd like to request

Currently, link prefetching on SSRd pages is limited to prefetching the js bundle needed for that page, while prefetching for SSG/ISR pages supports both code and prop prefetching. I am using SSR with a CDN / origin cache, and I want to be able to 'turn on' prop prefetching for SSR'd pages without having to opt in to SSG/ISR, which does not work for my usecase.

### Describe the solution you'd like

Add either a value to next.config.js that 'turns on' prop prefetching for ssr pages (maybe based on path patterns like headers, but I would be OK with turning it on globally). With this feature enabled, prefetching for SSRd pages would work exactly like prefetching for SSG/ISRd pages.

### Describe alternatives you've considered

I have tried rewriting my site to use ISR with fallback: blocking, and a very low 'revalidate' time, but I have multiple 'hosts' and therefore multiple ISR caches, and busting that cache on each host on each underlying data (content in a CMS) change is infeasible.

**Comment by nihakue** [comment]
Why was this moved from feature request to discussion? Is that expected? Did I fill in the feature request template incorrectly? @timneutkens

**Reply by timneutkens** [reply]
It's expected as per the issue template, all feature requests are moved to discussions "ideas" section. Unfortunately GitHub does not support templates for discussions yet otherwise we'd have you create them in discussions immediately.

**Comment by nihakue** [comment]
I've started working on a PR to add an experimental flag that makes prefetching work for SSR props by:

1. In `router.prefetch`, check isSsg, but in the second part of the ternary call `_getServerProps` instead of doing nothing
2. Turn `router.sdr` into a cache instead of an in-flight request cache

There are two main hurdles so far:

1. SSR props aren't cached by default, wheras SSG props are. This cache doesn't appear to be configurable, and also looks like it just fills until the user does a refresh where it's wiped completely. I don't know what the consequences of adding caching to SSR prop responses would be, but without doing that (or something like it) there's less value in prefetching because the browser will re-request the prefetched props on click. https://timkadlec.com/remembers/2020-06-17-prefetching-at-this-age/ has a good explanation of Chrome and Firefox's caching behavior for native prefetching, it seems sensible to do something similar (TTL of at least 5 minutes, regardless of caching headers). It might also be sensible to add a TTL of 5 minutes to the `sdc`, but less sure about that one.
2. I'm not sure how to rebuild the params/query part of the request for prefetched URLs. based on the url passed in to `prefetch`. I've been debugging the`router.change` method, and honestly I'm not sure where it's constructing the query. This is mostly just due to me being unfamiliar with the codebase, so any pointers would be appreciated. On that note, It seems slightly odd that various methods in `router` (namely `change` and `prefetch) have to manually construct the `dataHref`, maybe this could be extracted into a common method. I would have done this straight away, but I'm still wrapping my head around `routeInfo` and when it gets loaded / is available.

Any thoughts?
